reps/2023-8-18-ray-on-spark-autoscaling.md

```diff
@@ -0,0 +1,291 @@
+
+The second reason is ray-on-spark autoscaling only supports a very limited subset of cluster.yaml: e.g. only a single worker type, no docker, no ssh so it's easier and less confusing to use a restricted API.
+
+#### How to integrate Ray on spark NodeProvider backend with databricks spark cluster autoscaling feature ?
+
+See [Cluster size and autoscaling](https://docs.databricks.com/en/clusters/configure.html#autoscaling) and [Enhanced Autoscaling](https://docs.databricks.com/en/delta-live-tables/auto-scaling.html) for configuring databricks spark cluster autoscaling.
+
+Note that scaling up or down is triggered automatically. In Ray on spark NodeProvider
+backend, we just need to create or cancel spark jobs by demand, then databricks cluster
+will automatically scale up or down according to cluster resources utilization ratio.
+
+
+#### How to integrate Ray on spark NodeProvider backend with apache/spark application dynamic resource allocation on standalone mode spark cluster ?
+
+See [dynamic resources allocation](https://spark.apache.org/docs/latest/job-scheduling.html#dynamic-resource-allocation) for configuring spark application dynamic resource allocation.
+
+Note that scaling up or down is triggered automatically. In Ray on spark NodeProvider
+backend, we just need to create or cancel spark jobs by demand, then the spark application executors will automatically scale up or down according to this spark application resources utilization ratio.
+
+
+#### How to make `NodeProvider` backend support multiple Ray worker nodes running on the same virtual machine ?
+
+By default, `NodeProvider` implementation implement `internal_ip` and `external_ip` methods and convert `node_id` to IP, and different node must have different IP address,
+and Ray autoscaler internal code uses IP to track specific node status.
+
+But, for Ray on spark, one virtual machine (running a spark worker) might start multiple
+ray worker nodes. It breaks rules above, the workaround is, we can set `use_node_id_as_ip` configuration to `True` in autoscaler configuration file:
+
+```yaml
+provider:
+    type: ray_on_spark
+    use_node_id_as_ip: True
+    ...
+```
+
+and we use incremental number as the "fake node id" used in NodeProvider implementation code, and before creating Ray node process, setting an environment variable `NODE_ID_AS_RESOURCE` to be the "fake node id", then Ray autoscaler code can track
+these Ray nodes status correctly.
+See related logic code [here](https://github.com/ray-project/ray/blob/7a8b6a1b52488922fc27a1bc2a01d40f87d36af6/python/ray/autoscaler/_private/monitor.py#L304C34-L304C34) for reference.
+
+
+#### Shall we merge multiple Ray worker nodes into one spark job ?
+
+This is difficult. When we cancel a spark job, all spark tasks in the job are killed.
+But the `NodeProvider` interface might trigger `terminate_node` solely on specific
+Ray node.
```

Comment:
I feel this doesn't really answer the question of why we can't use the existing single spark job, one spark task per worker node model to support autoscaling.

Reply:
Updated